app/models/binary/Publication.scala

```diff
@@ -30,6 +37,46 @@ class PublicationService @Inject()()(implicit ec: ExecutionContext) {
         "created" -> p.created
       ))
 }
+class PublicationService @Inject()(dataSetService: DataSetService,
+                                   dataSetDAO: DataSetDAO,
+                                   dataStoreDAO: DataStoreDAO,
+                                   dataStoreService: DataStoreService,
+                                   tracingStoreDAO: TracingStoreDAO,
+                                   tracingStoreService: TracingStoreService,
+                                   annotationService: AnnotationService,
+                                   annotationDAO: AnnotationDAO)(implicit ec: ExecutionContext) {
+
+  def publicWritesWithDatasetsAndAnnotations(publication: Publication): Fox[JsObject] = {
+    implicit val ctx: DBAccessContext = GlobalAccessContext
+    for {
+      dataSets <- dataSetDAO.findAllByPublication(publication._id) ?~> "not found" ~> NOT_FOUND
+      annotations <- annotationDAO.findAllByPublication(publication._id) ?~> "not found" ~> NOT_FOUND
+      dataSetsJson <- Fox.serialCombined(dataSets)(d => dataSetService.publicWrites(d, None, None, None))
+      annotationsJson <- Fox.serialCombined(annotations) { annotation =>
+        for {
```

Comment:
I’d suggest to extract this into a separate function, either here, or in the annotation service. maybe `annotationWritesWithDataset`?